Wordcamp docker

vedanshujain opened the pull request:
Adds docker for provisioning WordCamp. This will create `.data` directory to persists MySQL. 

See added `readme.md` for more details.

vedanshujain committed:
WordCamp: Add docker setup to ease development flow

This is adds initial docker configuration for running WordCamp meta environment. Right now expectation is to have a MySQL running on host and configured in wp-config. Eventually MySQL will be added as part of docker configuration.

vedanshujain committed:
WordCamp: Add .data and .idea to gitignore to untrack local files

.data will contain local database files if docker is being used and .idea contains Jetbrans IDE config files. Both of these are local to dev environment, and there fore can be ignored.

vedanshujain committed:
WordCamp: Move docker config files to .docker/ for decluttering root.

This creates a `.docker` directory where all the config and provisioning files for docker environment can live therefore not cluttering the root folder.

vedanshujain committed:
WordCamp: Add MySQL service to docker for easier installation

This adds MySQL service inside docker so that there won't be any need to maintain separate MySQL instance for WordCamp related development.

vedanshujain committed:
WordCamp Docker: Add wp-config for docker installation.

This file adds docker specific wp-config file. Right now its stored in `.docker` directory so that it does not overwrite any `wp-config` file user may already have. This file is then mounted when docker starts.

vedanshujain committed:
WordCamp Docker: Add scripts to easy install plugins and themes

This script can be run manually once the container is loaded, to download and install all the themes and plugins that comes by default in WordCamp.

vedanshujain committed:
WordCamp Docker: Coding standards fixes

vedanshujain committed:
WC Docker: Create min files if they don't exist for failing `filemtime`

themes files are built as part of blocks output and are not included in git. These files not being present in initial run will cause issues with `filemtime` function.

vedanshujain committed:
WC Docker: Add readme for clear docker setup instructions.

Also use `wp-cli` instead of shell commands to install plugins and themes.

vedanshujain committed:
WC Docker: Add instructions to use custom database.

vedanshujain committed:
WC Docker: Force https for consistency and avoid mixed content

Earlier https was required but not forced, cause mixed content warning with some images. This patch will redirect all http requests to https.

vedanshujain committed:
WC Docker: Prevent overwriting `rest_route` if json_route is null

JSON REST API plugin is depreciated and will not be there for new installations. This patch adds a conditional check so that in cases `json_route` is not present, core's API handlers will handle the requests.
Request with `json_route` will anyways fail when JSON REST API plugin is not installed, although this should be rare because it should be hardcoded only in some old posts.

iandunn commented on .gitignore:
```diff
@@ -1 +1,3 @@
 ./public_html/wp-content/mu-plugins/blocks/assets/blocks.min*
+.data/
+.idea
```
`SHOULD` - I think `.idea` is similar to other things, like `.DS_Store`, `node_modules`, etc. It should be in people's `~/.gitignore_global`, instead of individual repos. It's specific to a dev's local system -- some devs may use PhpStorm while others use Sublime Text, Netbeans, etc; we wouldn't want to have to add all possible IDEs/editors. 

It's also in every repo that person works on, so it's better to just ignore it once in `.gitignore_global` instead of every individual repo.

vedanshujain replied:
Agree. Removed `.idea` in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .gitignore:
```diff
@@ -1 +1,3 @@
 ./public_html/wp-content/mu-plugins/blocks/assets/blocks.min*
+.data/
```
`MIGHT` - I would if that directory could be in `.docker/` instead, since it's a consequence of Docker, rather than being relevant to all local dev environments?

That'd keep the folder structure and this file less cluttered, and make it obvious where it comes from.

`SHOULD` - I think `mysql` or `database` would be a more descriptive/meaningful name.

vedanshujain replied:
Moved `.data` to `.docker/database` and updated refs in readme.md in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,55 @@
+FROM wordpress:5.0.3-fpm
+
+# Move WordPress files to Working Directory
+RUN mv /usr/src/wordpress /usr/src/public_html
```
`SHOULD` - It seems like it should be in `public_html/mu` to match production.

vedanshujain replied:
I am not in favor of doing this because it would require more non obvoius configuration, which I am not sure if we should do (atleast at this point). If you feel strongly about it though then I will open a Github issue to track it.

iandunn replied:
What's the extra config that's needed?

On the WP side, it's just defining `WP_CONTENT_DIR` and `WP_CONTENT_URL`. It doesn't seem like it'd need any custom config in Docker, unless `wordpress:5.0.3-fpm` is some kind of special package that doesn't let you config the path or something.

iandunn replied:
I also personally find it really annoying to have a 3rd-party external like WP mixed into the root of the site, and it can cause sometimes cause problems (like the recurring `update-root-files` fiasco on w.org).

I'm open to it if the alternatives are worse, though.

iandunn replied:
Er, I was thinking of configuring custom content folders, not installing WP in a subfolder. What's the config that's needed? I always install WP in a subfolder, so I might have Nginx config you can just copy/paste, etc.

vedanshujain replied:
Ah I see. Yeah we can move WordPress files to `/public_html/mu`. Will take care of this

vedanshujain replied:
Done in https://github.com/WordPress/wordcamp.org/pull/113/commits/9b89964fb5e1eaf28b8792cd54d3f9f0a937f79b

vedanshujain committed:
WC Docker: Replace conflicting nginx configuration

Replace nginx default conf because it conflicts with wordcamp's configuration on port 80

iandunn commented on readme.md:
```diff
@@ -0,0 +1,63 @@
+
+1. Using VVV. WordCamp.org is available as part of WordPress.org's meta network, and can be setup by following meta-environment setup guide: https://github.com/WordPress/meta-environment/blob/master/docs/install.md
+
+1. If you'd like to only setup WordCamp.org and would like to use Docker, you can use inbuilt docker provisioning. Please follow these steps
```
`SHOULD` - I think the docker-specific stuff should be modularize the docker into `.docker/README.md`, and then this item should just link to that. That keeps this file more digestable, and focused on being an overview, and avoids overwhelming people who aren't using Docker with irrelevant info.

Rather than adding this file now, though, I think it'd be better to do that as part of #2, since that has a lot of other stuff we probably want to include.

vedanshujain replied:
I would think Docker and VVV instructions should reside within same parent folder. Maybe same level pages on make/meta or inside a docs folder in git repo with readme linking to those. What do you think?

iandunn replied:
🤔 Having them inside the repo is probably best, so that people don't have to switch back and forth between a browser and IDE/terminal. Having it in the repo also gives us a rendered version on GitHub.com that we can link to from the handbooks.

I think it'd be good for `/README.md` to _mention_ the VVV and Docker instructions, and link to them, but having the full instructions in the main readme feels overwhelming and cluttered to me. I think it'll make it difficult for people to navigate that document and find what they need.

The VVV instructions should probably remain inside the Meta Environment repo, though, since that's where the provisioning code/database/etc is. It'd be good to mention/link to it from `/README.md`, though.

I like the idea of a `docs/` folder, but it doesn't seem like we have enough for that yet, since right now it's just this file. Maybe as part of #2 we could discuss moving some of our internal devdocs there, though? Even if we did move it there, I still think the best location for the Docker readme would be inside `.docker/`, but it'd definitely be good to link (or just symlink) to that from `docs/`.

I'm open to other ideas, though. Do you have any thoughts, @coreymckrill ?

vedanshujain replied:
I splitted readme.md in https://github.com/WordPress/wordcamp.org/pull/113/commits/db5db5a1d8c4db4cd3f90b0c97aa60795cfa729a . Please resolve this conversation if this is good enough for now.

iandunn replied:
👍

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,55 @@
+RUN cp ~/wordcamp.crt /etc/ssl/certs/wordcamp.crt
+RUN cp ~/wordcamp.key /etc/ssl/private/wordcamp.key
+
+# IAdd script to install plugin and themes
```
```suggestion
# Add script to install plugin and themes
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on readme.md:
```diff
@@ -0,0 +1,63 @@
+        
+        `/wp-admin` pages for these sites should now be accessible. Use `admin` as username and `password` as password to login.
+        
+        **Note:** `https` URL scheme must be used to visit these sites. Security exception will be required in first time run.
```
`MIGHT` - It'd be nice to add instructions for adding the cert to the browser, so you get a green link instead of yellow. That's not necessary, just a nice-to-have

vedanshujain replied:
Punting this for later. Because the process will actually be other way around. You would generate a cert on host first and then add it inside Docker. 
Keeping this conversation un-resolved for now, and will open a issue when we merge this.

iandunn replied:
👍

vedanshujain replied:
Raised https://github.com/WordPress/wordcamp.org/issues/121

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,55 @@
+# RUN sh install-plugin-theme.sh
+
+
+# Install wkhtmltopdf for camptix invoices. See https://stackoverflow.com/a/38336153/1845153
```
```suggestion
# Install `wkhtmltopdf` for WordCamp Docs, CampTix Invoices, etc. See https://stackoverflow.com/a/38336153/1845153.
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,55 @@
+RUN tar xvf wkhtmltox.tar.xz
+RUN mv wkhtmltox/bin/wkhtmlto* /usr/bin/
+
+# Install cli
```
```suggestion
# Install WP-CLI.
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,55 @@
+RUN mv wp-cli.phar /usr/local/bin/wp
+RUN apt-get install less -y
+
+# Install Mail catcher
```
```suggestion
# Install MailCatcher.
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/install-plugin-theme.sh:
```diff
@@ -0,0 +1,9 @@
+
+# Install plugins
+echo "Installing plugins"
+wp --allow-root --url=wordcamp.test plugin install akismet buddypress bbpress jetpack wp-multibyte-patch wordpress-importer polldaddy liveblog wp-super-cache custom-content-width camptix-network-tools email-post-changes tagregator supportflow camptix-pagseguro camptix-payfast-gateway camptix-trustpay camptix-trustcard camptix-mercadopago camptix-kdcpay-gateway campt-indian-payment-gateway camptix
```
`MUST` - CampTix should probably be cloned from GitHub, since we'll want the development version.

IIRC the Meta Environment provision script does that with a few others too.

vedanshujain replied:
I thought about this, but I guess responsibility of VCS for anything inside `wp-content` folder should be on host instead of Docker 🤔 . So if someone wants to develop on CampTix, they should ideally replace the directory inside `wp-content/plugin` folder. What do you think?

iandunn replied:
🤔 , I thought that Docker shared the folder between the host and container, so that they're essentially the same? Given that, I'm not sure why the responsibility would shift?

It seems like we'd just be adding an extra step that has to be done manually, rather than automating it.

vedanshujain replied:
Docker does shared folder, however VCS tools like Git or SVN are not guaranteed to be installed inside it.
I visualize Docker setup like a run time environment, so everything related to running the site should be inside the Docker, but source control should be outside it. For egs, someone can download the source code zip and be able to run Docker, without having to setup or log in to Git, SVN or any other tool which is not required to run it.
Since CampTix is also open source, there is no blocker to actually clone it from GitHub although I feel the setup is much cleaner because of this distinction, but that's just a preference.

iandunn replied:
🤔 yeah, I can see that point of view. I personally still lean towards the view that the practical goal is more important than maintaining purity.

Another way to archive the goal would be to make CampTix a Git submodule, but Git submodules are pretty awful. 

If there are any other steps like this, then maybe another solution would be to provide a script that the user could run post-installation in order to replace the downloaded code w/ a cloned repo, etc. Maybe not worth it if this is the only thing, though.

🤷‍♂ , not a big deal either way. Go with whatever feels best to you.

vedanshujain replied:
For now lets keep these separate. If people complain about this, then we will revisit and make the change. Resolving the conversation for now

iandunn commented on .docker/install-plugin-theme.sh:
```diff
@@ -0,0 +1,9 @@
+
+# Install themes
+echo "Installing themes"
+wp --allow-root --url=wordcamp.test theme install twentyten twentyeleven twentytwelve twentythirteen twentyfourteen
```
`MUST` - I think we'll need the rest of the Twenty* themes too. We can't rely on Core to bundle them in the future, because that changes over time.

vedanshujain replied:
This would add some non-neglible time to setup procedure. I added this because they were also getting added by meta environment, and TBH I wanted to install not more than 2 3 themes.

User can always add whatever themes they like on their own. What do you think?

iandunn replied:
I think the fundamental principle of development environments is that they should match production as closely as possible. Having a fast setup time is nice, but I think accurately representing production is far more important.

vedanshujain replied:
Added themes in https://github.com/WordPress/wordcamp.org/pull/113/commits/db214d69ebe62cfa4be3af4e8432092c90f72061

iandunn commented on .docker/install-plugin-theme.sh:
```diff
@@ -0,0 +1,9 @@
+
+# Install themes
+echo "Installing themes"
+wp --allow-root --url=wordcamp.test theme install twentyten twentyeleven twentytwelve twentythirteen twentyfourteen
```
`MUST` - Also `p2`

vedanshujain replied:
Same as https://github.com/WordPress/wordcamp.org/pull/113#discussion_r281663564

vedanshujain replied:
Added themes in https://github.com/WordPress/wordcamp.org/pull/113/commits/db214d69ebe62cfa4be3af4e8432092c90f72061

iandunn commented on .docker/nginx.conf:
```diff
@@ -0,0 +1,51 @@
+# Upstream to abstract backend connection(s) for php
```
```suggestion
# Upstream to abstract backend connection(s) for PHP.
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/wp-config.php:
```diff
@@ -0,0 +1,129 @@
+<?php
+/**
+ * The base configuration for WordPress
```
`SHOULD` - I think all the comments can be removed, they're intended for new users and just clutter the file. Anything non-obvious should be documented, though.

vedanshujain replied:
Removed redundant comments in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/wp-config.php:
```diff
@@ -0,0 +1,129 @@
+<?php
```
`SHOULD` - The Meta Environment version might be good to look at to see if it has anything that's missing from here.

vedanshujain replied:
Synced and added some missing configs in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on .docker/wp-config.php:
```diff
@@ -0,0 +1,129 @@
+define( 'JETPACK_DEV_DEBUG', true);
+define( 'SAVEQUERIES', true );
+define( 'SCRIPT_DEBUG', true );
+define( 'MULTISITE', true );
```
`MUST` - Run the linter to align these, and catch other deviations from the coding standards.

vedanshujain replied:
Aligned now. Although interestingly this does not get flagged by the linter setup to run on PRs. Is this flagged by your linter?

iandunn replied:
Ah, you're right, the linter doesn't flag that yet. I'll add a `todo` comment to it.

vedanshujain replied:
Created an issue to track it. Resolving the conversation now. https://github.com/WordPress/wordcamp.org/issues/120

iandunn commented on .docker/wp-config.php:
```diff
@@ -0,0 +1,129 @@
+ */
+define('WP_DEBUG', true);
+
+/* That's all, stop editing! Happy blogging. */
```
`SHOULD` - The only stuff past this line should be defining `ABSPATH` and including `wp-settings`

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+version: "3.2"
```
`SHOULD` - Can this live in `.docker/` ?

vedanshujain replied:
It can but then folks would have to go inside a folder starting with `.` to start a docker. This might be annoying for several windows programs like Windows. IMO its fine to clutter the root with this file.

iandunn replied:
That's a good point.

🤔 

It sounds like we could get around that by [setting the `COMPOSE_FILE` environment variable](https://docs.docker.com/compose/reference/envvars/#compose_file):

`COMPOSE_FILE=.docker/docker-compose.yml:other/paths/the/user/wants`

We'd have to add that to the README, though, so it does add a bit of extra work for the user, and potential confusion.

I'm probably just being too pedantic, so the root folder is fine with me unless @coreymckrill would also prefer it in a subdir.

vedanshujain replied:
I still lean towards not having to create an environment variable. I have not tested it out, but it looks like `COMPOSE_FILE` is used to specify multiple compose files, instead of defining a path. Path can simply be specified while executing `docker-compose` command. 

I am resolving this conversation now, but please re-open this if you feel we should rethink.

iandunn commented on readme.md:
```diff
@@ -0,0 +1,63 @@
+        
+        `/wp-admin` pages for these sites should now be accessible. Use `admin` as username and `password` as password to login.
+        
+        **Note:** `https` URL scheme must be used to visit these sites. Security exception will be required in first time run.
```
`MUST` - Do `HTTP` URLs redirect to `HTTPS`?

vedanshujain replied:
Yes they do. See https://github.com/WordPress/wordcamp.org/pull/113/files#diff-821cdce799b2a90c846089e20e997ff2R8

iandunn replied:
👍

iandunn commented on readme.md:
```diff
@@ -0,0 +1,63 @@
+        
+        **Note:** `https` URL scheme must be used to visit these sites. Security exception will be required in first time run.
+        
+    1. (Optional) Basic installation comes without themes and with only very basic plugins. Add plugins and themes which are installed in wordcamp.org if needed like so:
```
`MUST` - Can this be automated as part of installation?

vedanshujain replied:
Unfortunately I was not able to find a non complicated solution to do this. Main reason being that we mount the `wp-content` directory from host filesystem, which is going to overwrite Docker image's `wp-content` folder.

iandunn replied:
🤔 I'm confused, I would think they'd be the same folder, so any changes within the container would be synced to the host. Is that not the case?

vedanshujain replied:
Yup, so during the first time docker build process, a container image will be build. But while starting up the container, we will mount the host's `wp-content` directory inside that image (overwriting image's wp-content folder) so that files are synced. That's why if during the build process we install the plugins and themes, they will get overwritten. 

We can solve this by not mounting `wp-content` but instead sub-folders one by one, but that will make the config very big and hard to keep updated. Let me also take a look to see if there are other solutions, but so far this looks like to be the case.

iandunn replied:
🤔 

huh. Maybe this is another case where we'd want [a script that the user runs on the host after installation](https://github.com/WordPress/wordcamp.org/pull/113#discussion_r284009060)?

Then it seems like we couldn't need to install anything during the initial build.

vedanshujain replied:
We already have a script for installing plugins and themes in `install-pluign-theme.sh`. Perhaps we can mark this step as mandatory instead of optional. What do you think?

iandunn replied:
That sounds good to me, since they won't really have a completely functioning environment until they do that.

vedanshujain replied:
Done in https://github.com/WordPress/wordcamp.org/pull/113/commits/8e679936b8e51e534ca63ea91ea26639384690de

iandunn commented on readme.md:
```diff
@@ -0,0 +1,63 @@
+            
+        1. Activate plugin or apply theme as needed.
+        
+    1. (Optional) By default, docker will start with data defined in `.docker/wordcamp_dev.sql` and changes to data will be persisted across runs. To start with different database, delete `.data` directory and replace the `.docker/wordcamp_dev.sql` file and run `docker-compose up --build` again.
```
```suggestion
    1. (Optional) By default, docker will start with data defined in `.docker/wordcamp_dev.sql` and changes to data will be persisted across runs in `.data/`. To start with different database, delete `.data` directory and replace the `.docker/wordcamp_dev.sql` file and run `docker-compose up --build` again.
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn commented on readme.md:
```diff
@@ -0,0 +1,63 @@
+    
+    After first time provisioning, docker can be started by using `docker-compose up` command from inside the directory. 
+        
+To contribute, you can send pull requests to this repo, or add patches to https://meta.trac.wordpress.org/.
```
```suggestion
To contribute, you can send [pull requests on GitHub](https://github.com/WordPress/wordcamp.org/), or add [patches on Meta Trac](https://make.wordpress.org/meta/handbook/about/get-involved/learn-how-to-contribute-code/#share-your-new-code).
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/1e3bc2e5e0ebcc4643db21c0d98a69b94a51b6c8

iandunn reviewed (changes requested):
Looks good overall. I left a few suggestions.

I didn't spend any time testing it, since it sounds like @coreymckrill is eager to try an alternative to VVV.

vedanshujain committed:
WC Campsite 2017: Code review fixes. See desc:

1. Removed redundant comments from `wp-config.php`
2. Moved `.data` to inside `.docker/database`
3. Fixed typos and capitalization
4. Added any missing settings from https://github.com/WordPress/meta-environment/blob/master/wordcamp.test/provision/wp-config.php

vedanshujain committed:
WC Docker: Install libext6 because`wkhtmltopdf` errors without this lib.

vedanshujain committed:
WC Docker: De clutter main `readme.md` for readability

This patch splits `readme.md` into `.docker/readme.md` and `readme.md` for better readability.

vedanshujain committed:
WC Docker: Add `twenty*` themes to match production closely

We already have default theme installed in production, so it makes sense to install them in Docker as well.

vedanshujain committed:
WC Docker: Move WordPress files to `./mu` to match production.

In production environment, we have WordPress files in `public_html/mu` directory, so it makes sense to have them in same location in Docker environment as well.

This also required config change in `nginx` so that static files in `wp-content` are served properly given that root is changed.

This patch also has a minor fixup of http->https.

vedanshujain committed:
WC Docker: Add default config for WP-CLI to work with `mu` path change

In an earlier commit, we changed default WordPress installation location to `/mu` folder. This patch adds default argument in WP-CLI to work with that change.

vedanshujain committed:
WC Docker: Typo fixups in readme.md

vedanshujain committed:
WC Docker: Build from php-7.2-fpm to allow easy switching WP version

We will clone the WordPress's Github repo. This allow to change the current WordPress version by just checking out the required branch (like `git checkout 5.2-branch`).

Also added some steps to install required dependencies.

vedanshujain committed:
WC Docker: Plugin and themes install is mandatory to reflect production

This marks the step to install plugins and themes as mandatory to reflect production. Earlier this was optional, but since we have all these in production anyway, it makes sense to also include them here.

Also, some small readme changes for blocks installation.

vedanshujain committed:
WC Docker: Some more readme enhancements for clearer instructions

vedanshujain committed:
WC Docker: Change block's min file paths after folder restructuring

We recently changed folder structure for blocks, this patch changes the file path in docker-compose.yml to account for that.

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,120 @@
+
+# Unfortunately themes and plugins cannot be installed as part of docker setup and have to be run manually.
+# This is because everything in wp-content folder will be overwritten when we mount this directory from volume.
+# RUN sh install-plugin-theme.sh
```
`SHOULD` - It doesn't seem like we need this comment here anymore, and maybe not the `ADD` command above, since they're artifacts of when it was run automatically.

Now that it's run manually, it seems like it makes more sense for the script to just be mentioned in the `README`, and then add a comment inside the script itself that documents why it can't be run automatically during `compose`.

vedanshujain replied:
I moved the comment over to `readme.md` and removed the `RUN` command. Although, we still need `ADD` command to actually push the file inside docker environment.

vedanshujain replied:
^ this was done in https://github.com/WordPress/wordcamp.org/pull/113/commits/41867e7168934f7be3093a8ade78a9c1d3604520

iandunn replied:
👍

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,120 @@
+
+COPY wp-config.php /usr/src/public_html/wp-config.php
+
+# Replace nginx default conf because it conflicts with wordcamp.cong
```
```suggestion
# Replace nginx default conf because it conflicts with wordcamp.conf
```

`MUST`

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/41867e7168934f7be3093a8ade78a9c1d3604520

iandunn commented on .docker/Dockerfile:
```diff
@@ -0,0 +1,120 @@
+RUN mv /etc/nginx/sites-enabled/wordcamp.conf /etc/nginx/sites-enabled/default
+
+CMD tail -f /dev/null
+
```
```suggestion

RUN echo "\n\nSuccess: The WordCamp container is now running.\n"
RUN echo "You can visit https://2014.content.wordcamp.org/ to view a sample WordCamp site. You can log in with username 'admin' and password 'password'.\n"
RUN echo "You can also visit http://localhost:1080/ to open MailCatcher."

```

`MUST` - At first I got tripped up because I didn't realize that `docker-compose` up was supposed to run indefinitely. I expected it to finish and return me to the shell, similar to a `vagrant up`, but it just sat there after starting MySQL for several minutes before I got it. This will help avoid that, and also give the user some necessary information.

Alternatively, it'd be cool if instead of showing the Nginx access log, if the terminal instead automatically entered a shell inside the container, so that you don't have to do that manually. If we don't do that, we'll probably want a quick way to enter it without having to lookup the current name and type a command. Maybe just `ssh wordcamp.test` via `~/.ssh/config` or something?

vedanshujain replied:
I have edited Docker's `readme.md` and changed some of the provisioning command.
`docker-compose up --build -d` will exit when provisioning is done and docker instance will keep on running in background. 

Also `docker-compose exec wordcamp.test bash` can be used to go inside the shell. See https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60

Please resolve the conversation if you feel that this addressed your issues.

iandunn replied:
That commit looks great!

iandunn commented on .docker/install-plugin-theme.sh:
```diff
@@ -0,0 +1,9 @@
+
+# Install themes
+echo "Installing themes"
+wp --allow-root theme install twentyten twentyeleven twentytwelve twentythirteen twentyfourteen twentyfifteen twentysixteen twentyseventeen twentyeighteen p2
```
```suggestion
wp --allow-root theme install twentyten twentyeleven twentytwelve twentythirteen twentyfourteen twentyfifteen twentysixteen twentyseventeen twentynineteen p2
```

`MUST` - There wasn't a `twentyeighteen` theme, we skipped that year, but there is a `twentynineteen`

```sh
Installing Twenty Seventeen (2.2)
Downloading installation package from https://downloads.wordpress.org/theme/twentyseventeen.2.2.zip...
The authenticity of twentyseventeen.2.2.zip could not be verified as no signature was found.
Unpacking the package...
Installing the theme...
Theme installed successfully.
Warning: twentyeighteen: Theme not found
Warning: The 'twentyeighteen' theme could not be found.
***************** ^
Installing P2 (1.5.8)
Downloading installation package from https://downloads.wordpress.org/theme/p2.1.5.8.zip...
The authenticity of p2.1.5.8.zip could not be verified as no signature was found.
Unpacking the package...
Installing the theme...
Theme installed successfully.
Error: Only installed 9 of 10 themes.
***************** ^
```

vedanshujain replied:
Done in https://github.com/WordPress/wordcamp.org/pull/113/commits/41867e7168934f7be3093a8ade78a9c1d3604520

iandunn commented on .docker/nginx.conf:
```diff
@@ -0,0 +1,57 @@
+}
+
+server {
+    listen 80 default_server;
```
`MIGHT` - I wonder if it'd be better to use non-standard ports for this and `443`? I have a local Nginx install that I leave running 24/7, and it has a bunch of other sites in addition to WordCamp.test. It'd be really inconvenient to have to stop this container, and restart the local Nginx/MySQL/PHP servers every time I want to change to another site.

It seems like that might be common, but if I'm the only one, though, maybe I can work out another solution. It might be normal for people to only run one container at a time?

vedanshujain replied:
So this configuration will actually go into `docker-compose.yml:11` where you can choose a different port on your host machine to map to port 80 of Docker's webserver. For egs, we map host's 3307 to MySQL Docker's 3306 to avoid conflict. 

That said, using any port other than 80 might not work because lots of image path are hardcoded without port number in the database. So we would have to clean that first, which should ideally be a different PR.

You can comment out the line in `docker-compose.yml` in your local setup if docker-setup conflicts with your local one.

iandunn replied:
🤔 

Yeah, let's just go with this for now, and I can play around with it later and maybe iterate.

I wonder if there's a way to do an "local" config that overrides parts of `docker-compose.yml`, similar to how you can have `phpcs.xml.dist` committed to version control, and then an ignored `phpcs.xml` file that takes precedence.

iandunn commented on .docker/php-fpm.conf:
```diff
@@ -0,0 +1,6 @@
+[www]
+php_flag[display_errors] = on
+php_admin_value[error_log] = /var/log/fpm-php.www.log
+php_admin_flag[log_errors] = on
```
```suggestion
php_admin_flag[log_errors] = on
php_admin_flag[post_max_size] = 100M
php_admin_flag[upload_max_filesize] = 100M
```

`SHOULD` - Now that this is in a separate file, the `if you'd like too...` seems like an artifact leftover from when this sentence was next to one referring to the Meta Environment. If they're loading this file, we can assume they just want the Docker process.

vedanshujain replied:
Added in https://github.com/WordPress/wordcamp.org/pull/113/commits/936ea89eb0aaaedcc90dda6925af03fdac01ad41

iandunn commented on .docker/readme.md:
````diff
@@ -0,0 +1,55 @@
+    git clone git@github.com:WordPress/wordcamp.org.git
+    ```
+
+1. Change into the directory, and run docker compose command. This could take some time depending upon the internet speed.
````
```suggestion
1. Change into the directory, and run Docker's `compose` command. This could take some time depending upon the speed of your Internet connection.
```

`SHOULD` (or something similar)

vedanshujain replied:
Done in https://github.com/WordPress/wordcamp.org/pull/113/commits/41867e7168934f7be3093a8ade78a9c1d3604520

iandunn commented on .docker/readme.md:
````diff
@@ -0,0 +1,55 @@
+    ```
+    cd wordcamp.org && docker-compose up --build
+    ```
+     This will also start the docker environment for WordCamp.org development.
````
```suggestion
     This will also start the docker environment for WordCamp.org development. This command will continue running as long as the environment is running.
```

`SHOULD` - At first I got tripped up because I didn't realize that that command was supposed to run indefinitely. I expected it to finish and return me to the shell, similar to a `vagrant up`, but it just sat there after starting MySQL for several minutes before I got it.

`MIGHT` - It'd also be nice to provide a way to just load the container when the user logins in to their OS account, rather than having to manually start it and leave it running in a terminal which then can't be used for anything else.

vedanshujain replied:
Default provisioning command are now changed in https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60 to exit when its complete, instances will keep on running in background. 

That command can also be now added to host's startup script to keep the Docker service running.

iandunn replied:
👍

iandunn replied:
> That command can also be now added to host's startup script to keep the Docker service running.

I'll look into that when I make the switch, and then add instructions to the README.

iandunn commented on .docker/readme.md:
````diff
@@ -0,0 +1,55 @@
+    127.0.0.1	2014.new-site.wordcamp.test
+    ```
+    
+    `/wp-admin` pages for these sites should now be accessible. Use `admin` as username and `password` as password to login.
````
```suggestion
    `/wp-admin` pages for these sites should now be accessible. Use `admin` as username and `password` as password to login. Front end pages will not be accessible until you complete the remaining steps.
```

`SHOULD`

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60

iandunn commented on .docker/readme.md:
```diff
@@ -0,0 +1,55 @@
+    
+    **Note:** `https` URL scheme must be used to visit these sites. Security exception will be required in first time run.
+    
+1. Basic installation comes without themes and with only very basic plugins. Add plugins and themes which are installed in wordcamp.org if needed like so:
```
```suggestion
The installation doesn't have any 3rd-party plugins or themes yet, you must add them like so:
```

`SHOULD` - it seems like the previous explanation is an artifact from before this step became required.

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60

iandunn commented on .docker/readme.md:
```diff
@@ -0,0 +1,55 @@
+    
+1. Basic installation comes without themes and with only very basic plugins. Add plugins and themes which are installed in wordcamp.org if needed like so:
+
+    1. Find the docker which is running WordCamp. It will start with `wordcamporg_wordcamp.test` and will most likely be `wordcamporg_wordcamp.test_1`
```
```suggestion
     1. Find the WordCamp container by running `docker ps`. It will be named similar to `wordcamp-docker_wordcamp.test_1`
```

`SHOULD` - this will help them find the exact name quickly if they're new to docker

vedanshujain replied:
Changed the default scripts in https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60. 

This can be done by using `docker-compose exec wordcamp.test bash`

iandunn replied:
Nice! That's much better.

iandunn commented on .docker/readme.md:
```diff
@@ -0,0 +1,55 @@
+    
+    1. Go inside this docker:
+    
+       **Note:** You might need to change `wordcamporg_wordcamp.test_1` to whatever is docker name in your setup.
```
```suggestion
```

`SHOULD` - This probably isn't needed if we provide them `docker ps` above

vedanshujain replied:
Changed in https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60

iandunn commented on .docker/readme.md:
````diff
@@ -0,0 +1,55 @@
+    
+    1. Run predefined command for installing plugins and themes. This could take some time depending on your internet connection:
+        ```bash
+           sh install-plugin-theme.sh
````
`MIGHT` - I'm getting some warnings:

```sh
> docker exec -it wordcamp-docker_wordcamp.test_1 bash
root@878d69d889f9:/usr/src/public_html# sh install-plugin-theme.sh
Installing plugins
Installing Akismet Anti-Spam (4.1.2)
Downloading installation package from https://downloads.wordpress.org/plugin/akismet.4.1.2.zip...
The authenticity of akismet.4.1.2.zip could not be verified as no signature was found.
************ ^
Unpacking the package...
Installing the plugin...
Plugin installed successfully.
```

I've never seen that before. It could be related to the 5.2 package signing stuff? Not sure.

vedanshujain replied:
🤔 Looks like it.

iandunn replied:
👍 Probably nothing for us to worry about here, then.

iandunn commented on .docker/readme.md:
````diff
@@ -0,0 +1,55 @@
+           sh install-plugin-theme.sh
+        ```
+        
+    1. Activate plugin or apply theme as needed.
````
`MUST` - Shouldn't the default DB already have them activated?

vedanshujain replied:
It is. removed this line. https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60

iandunn replied:
👍

iandunn commented on .docker/readme.md:
```diff
@@ -0,0 +1,55 @@
+
+After first time provisioning, docker can be started by using `docker-compose up` command from inside the directory. 
+
+Once the Docker instance has started, you can visit [2014.content.wordcamp.org](2014.content.wordcamp.org) to view a sample WordCamp site. WordCamp central would be [central.wordcamp.test](central.wordcamp.test). You can also visit [localhost:1080](localhost:1080) to view the MailCatcher dashboard.
```
```suggestion
Once the Docker instance has started, you can visit [2014.content.wordcamp.org](2014.content.wordcamp.org) to view a sample WordCamp site. WordCamp central would be [central.wordcamp.test](central.wordcamp.test). You can also visit [localhost:1080](localhost:1080) to view the MailCatcher dashboard.

To stop the WordCamp container, press `ctrl + c` in the terminal window that's running the `docker-compose` command.
```

`MUST` - Otherwise new Docker users won't know how to close it. I'm not sure if `ctrl-c` is the recommended way, or if there's something better that shuts it down gracefully.

vedanshujain replied:
Instructions are now changed, so this is not a problem anymore because `compose` command will automatically exit using `-d` flag. To stop the container use `docker-compose stop`. This is documented in https://github.com/WordPress/wordcamp.org/pull/113/commits/ac11b763700e157d761baedda4253bf3722d6e60

iandunn commented on .docker/install-plugin-theme.sh:
```diff
@@ -0,0 +1,9 @@
+
+# Install plugins
+echo "Installing plugins"
+wp --allow-root plugin install akismet buddypress bbpress jetpack wp-multibyte-patch wordpress-importer polldaddy liveblog wp-super-cache custom-content-width camptix-network-tools email-post-changes tagregator supportflow camptix-pagseguro camptix-payfast-gateway camptix-trustpay camptix-trustcard camptix-mercadopago camptix-kdcpay-gateway campt-indian-payment-gateway camptix
```
```suggestion
wp --allow-root plugin install akismet buddypress bbpress  classic-editor jetpack wp-multibyte-patch wordpress-importer polldaddy liveblog wp-super-cache custom-content-width camptix-network-tools email-post-changes tagregator supportflow camptix-pagseguro camptix-payfast-gateway camptix-trustpay camptix-trustcard camptix-mercadopago camptix-kdcpay-gateway campt-indian-payment-gateway camptix
```

`MUST` - I'm seeing this at https://wordcamp.test/wp-admin/network/upgrade.php

```
The following plugins are missing:

    classic-editor/classic-editor.php
    wporg-profiles-wp-activity-notifier/wporg-profiles-wp-activity-notifier.php
    json-rest-api/plugin.php

Please let a WordCamp.org developer know about this message.
```

`wporg-profiles-wp-activity-notifier` comes from the Meta repo. `json-rest-api` can be left out since we'll remove it soon anyway.

vedanshujain replied:
Added `classic-editor` in https://github.com/WordPress/wordcamp.org/pull/113/commits/41867e7168934f7be3093a8ade78a9c1d3604520

iandunn commented on .docker/readme.md:
```diff
@@ -0,0 +1,55 @@
+If you'd like to only setup WordCamp.org and would like to use Docker, you can use inbuilt docker provisioning. Please follow these steps
```
```suggestion
Follow these steps to setup WordCamp.org using [Docker](https://www.docker.com/).
```

`SHOULD` - Now that this is in a separate file, the if you'd like too... seems like an artifact leftover from when this sentence was next to one referring to the Meta Environment. If they're loading this file, we can assume they just want the Docker process.

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/41867e7168934f7be3093a8ade78a9c1d3604520

iandunn commented on .docker/readme.md:
```diff
@@ -0,0 +1,55 @@
+    
+1. By default, docker will start with data defined in `.docker/wordcamp_dev.sql` and changes to data will be persisted across runs in `.docker/database`. To start with different database, delete `.docker/database` directory and replace the `.docker/wordcamp_dev.sql` file and run `docker-compose up --build` again.
+
+1. Note that if you want to work on WordCamp blocks, [you would have to install all the node dependencies](../public_html/wp-content/mu-plugins/blocks/readme.md). This can be done either inside, or even from outside the Docker.
```
`SHOULD` - can this be automated in the install plugin script? If so that should maybe be renamed to `post-compose.sh` or something.

There's probably some other stuff that needs to be built, like the Application Tracker.

vedanshujain replied:
Having a separate that automatically does all this is a good idea, but perhaps it should not be part of this PR, because it can also be used in VVV (or any other) setup. I will open an issue to track it before merging. What do you think?

iandunn replied:
👍

vedanshujain replied:
Raised https://github.com/WordPress/wordcamp.org/issues/122

iandunn commented on .docker/wp-cli.yml:
```diff
@@ -0,0 +1,2 @@
+path: /usr/src/public_html/mu
+url: wordcamp.test
```
```suggestion
```

`MIGHT` - We can't use that on sandboxes/production, so it might be best to just leave it out. Alternatively, we could `gitignore` this file? But it would be nice to have one on production too, so maybe the script that pushes to the SVN repo could swap `.test` for `.org` or something?

vedanshujain replied:
Its added to `public_html` during Docker compose, but otherwise its always inside `.docker` folder. So it should not cause an issue on sandboxes/production. What do you think?

iandunn replied:
👍

iandunn commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+version: "3.2"
+services:
+  wordcamp.test:
```
`MIGHT` - Can we use tabs for indentation instead of spaces, to be consistent? IIRC YAML allows that.

vedanshujain replied:
Unfortunately, [YAML does not allow tabs](https://yaml.org/faq.html).

iandunn replied:
huh, ok :)

New issue: from that FAQ it sounds like maybe we should use `.yaml` instead of `.yml`, but not a big deal.

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/113/commits/0a485e31005fb66c47a0c5c1bfa51ad6a6ca680e

iandunn commented on docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+      - "1080:1080"
+    command: >
+      bash -c "touch /usr/src/public_html/wp-content/mu-plugins/blocks/build/blocks.min.js &&
+               touch /usr/src/public_html/wp-content/mu-plugins/blocks/build/blocks.min.css &&
```
`MIGHT`  - This won't be necessary if we automate the building of this and other things like the Application Tracker

vedanshujain replied:
Depends upon resolution of https://github.com/WordPress/wordcamp.org/pull/113#discussion_r285101877

iandunn commented on public_html/wp-content/mu-plugins/wcorg-json-api.php:
```diff
@@ -361,9 +361,11 @@ function wcorg_json_v2_compat( $request ) {
-	$request->query_vars['rest_route'] = $request->query_vars['json_route'];
-	unset( $request->query_vars['json_route'] );
-	$request->matched_query = preg_replace( '#^json_route=(.+)$#', 'rest_route=$1', $request->matched_query );
+	if ( isset( $request->query_vars['json_route'] ) ) {
```
`MUST` - Why is this change necessary? Because the `json-rest-api` plugin isn't available in the w.org repo anymore? Do the v1 and v2 endpoints still work on your sandbox with this change?

vedanshujain replied:
Yup, thats because `json-rest-api` plugin isn't available in w.org repo. I tested out few `v2` endpoints, and some `json_route` endpoints like `?json_route=/posts/types/page` and they seem to work.

Is there anything specific you have in mind that we should test?

iandunn replied:
Nope, that sounds good to me. I think this should be a separate commit from the other Docker stuff, and the commit message should document that it's because the plugin isn't in the repo (publicly) anymore. Looks good though :)

iandunn reviewed (changes requested):
Great job, kudos!

There's some smaller things I think should change, but nothing big.

vedanshujain committed:
WC Docker: Changes in `readme.md` file for clear and easy instructions.

This patch documents multiple `docker-compose` shortcuts that will be frequently used.

vedanshujain committed:
WC Docker: Code review fixes. See desc:

1. Language changes in `readme.md`
2. Added `twentynineteen`, `classic-editor` to required themes and plugins. Also clarified few instructions
3. Removed `IS_WORDCAMP_NETWORK` because that does not seem to be used
4. Fixed some typos

vedanshujain committed:
WC Docker: Add post and file `max-size` conf to match production.

iandunn reviewed (approved):
Great work!

There're a few unresolved threads but they're not blockers.

iandunn commented:
`MUST` - Oh, one more thing I just noticed:

```sh
> git stat
?? .docker/database/
?? public_html/wp-content/plugins/akismet/
?? public_html/wp-content/plugins/bbpress/
?? public_html/wp-content/plugins/buddypress/
?? public_html/wp-content/plugins/campt-indian-payment-gateway/
?? public_html/wp-content/plugins/camptix-kdcpay-gateway/
?? public_html/wp-content/plugins/camptix-mercadopago/
?? public_html/wp-content/plugins/camptix-pagseguro/
?? public_html/wp-content/plugins/camptix-payfast-gateway/
?? public_html/wp-content/plugins/camptix-trustcard/
?? public_html/wp-content/plugins/camptix-trustpay/
?? public_html/wp-content/plugins/camptix/
?? public_html/wp-content/plugins/custom-content-width/
?? public_html/wp-content/plugins/email-post-changes/
?? public_html/wp-content/plugins/jetpack/
?? public_html/wp-content/plugins/liveblog/
?? public_html/wp-content/plugins/polldaddy/
?? public_html/wp-content/plugins/supportflow/
?? public_html/wp-content/plugins/tagregator/
?? public_html/wp-content/plugins/wordpress-importer/
?? public_html/wp-content/plugins/wp-multibyte-patch/
?? public_html/wp-content/plugins/wp-super-cache/
?? public_html/wp-content/themes/p2/
?? public_html/wp-content/themes/twentyeleven/
?? public_html/wp-content/themes/twentyfifteen/
?? public_html/wp-content/themes/twentyfourteen/
?? public_html/wp-content/themes/twentyseventeen/
?? public_html/wp-content/themes/twentysixteen/
?? public_html/wp-content/themes/twentyten/
?? public_html/wp-content/themes/twentythirteen/
?? public_html/wp-content/themes/twentytwelve/
```

Our "svn to git migration" doc pushes the 3rd-party plugin/themes to `.git/info/exclude`, but I feel like we can probably go ahead and just add them to `.gitignore` now.

Seems like `.docker/database/` can be `.gitignore`'d too.

vedanshujain committed:
WC Docker: Use `yaml` extension to follow standard

vedanshujain committed:
WC Docker: Add pre-installed themes and plugins to gitignore.

These were added to `.git/info/exclude` in `git-svn` setup, now that we are moving to git, it makes sense to add in `.gitignore` as well.

Note that these are managed by `svn:external` on production.

vedanshujain commented:
@iandunn Added plugins and themes to .gitignore

vedanshujain commented:
Merged in r8808-r8836
